File: ex328.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}



class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null)    return head;
        
        ListNode head2 = head.next;
        ListNode n1 = head, n2 = head2;
        ListNode next1, next2;
        
        while (n2 != null) {
            next1 = n1.next.next;
            n1.next = next1;
            if (next1 != null)    n1 = next1;
            
            if (n2.next == null)    next2 = n2.next;
            else    next2 = n2.next.next;
            n2.next = next2;
            n2 = next2;
            
        }
        n1.next = head2;
        
        return head;
    }
}

File: code/ex328.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        ListNode* evenHead = new ListNode(0);
        ListNode* oddHead = new ListNode(0);
        int i = 0;
        ListNode* p = head;
        ListNode* even = evenHead;
        ListNode* odd = oddHead;
        while (p != NULL) {
            if (i % 2 == 0) {
                even->next = p;
                p = p->next;
                even = even->next;
                even->next = NULL;
            } else {
                odd->next = p;
                p = p->next;
                odd = odd->next;
                odd->next = NULL;
            }
            i++;
        }
        even->next = oddHead->next;
        return evenHead->next;
    }
};